src/scene/shader-pass.js

```diff
@@ -1,110 +1,148 @@
-    SHADER_SHADOW, SHADOW_COUNT, LIGHTTYPE_COUNT,
-    SHADERTYPE_FORWARD, SHADERTYPE_DEPTH, SHADERTYPE_PICK, SHADERTYPE_SHADOW
+    SHADER_FORWARD, SHADER_FORWARDHDR, SHADER_DEPTH, SHADER_PICK, SHADER_SHADOW
 } from './constants.js';
 
+import { DeviceCache } from '../platform/graphics/device-cache.js';
+
+// device cache storing shader pass data per device
+const shaderPassDeviceCache = new DeviceCache();
+
 /**
- * A pure static utility class, responsible for math operations on the shader pass constants.
+ * Info about a shader pass. Shader pass is represented by a unique index and a name, and the
+ * index is used to access the shader required for the pass, from an array stored in the
+ * material or mesh instance.
  *
  * @ignore
  */
-class ShaderPass {
-    /**
-     * Returns the shader type given the shader pass.
-     *
-     * @param {number} shaderPass - The shader pass.
-     * @returns {string} - The shader type.
-     */
-    static getType(shaderPass) {
-        switch (shaderPass) {
-            case SHADER_FORWARD:
-            case SHADER_FORWARDHDR:
-                return SHADERTYPE_FORWARD;
-            case SHADER_DEPTH:
-                return SHADERTYPE_DEPTH;
-            case SHADER_PICK:
-                return SHADERTYPE_PICK;
-            default:
-                return (shaderPass >= SHADER_SHADOW && shaderPass < SHADER_SHADOW + SHADOW_COUNT * LIGHTTYPE_COUNT) ? SHADERTYPE_SHADOW : SHADERTYPE_FORWARD;
+class ShaderPassInfo {
+    /** @type {number} */
+    index;
+
+    /** @type {string} */
+    name;
+
+    /** @type {string} */
+    shaderDefine;
+
+    constructor(name, index, options = {}) {
+
+        Debug.assert(/^[a-zA-Z][_a-zA-Z0-9]*$/.test(name), `ShaderPass name can only contain letters, numbers and underscores and start with a letter: ${name}`);
+
+        this.name = name;
+        this.index = index;
+
+        // assign options as properties to this object
+        Object.assign(this, options);
+
+        this.initShaderDefine();
+    }
+
+    initShaderDefine() {
+
+        let keyword;
+        if (this.isShadow) {
+            keyword = 'SHADOW';
+        } else if (this.isForward) {
+            keyword = 'FORWARD';
+        } else if (this.index === SHADER_DEPTH) {
+            keyword = 'DEPTH';
+        } else if (this.index === SHADER_PICK) {
+            keyword = 'PICK';
         }
+
+        keyword ??= this.name.toUpperCase();
+        this.shaderDefine = `#define ${keyword}_PASS\n`;
     }
+}
 
+/**
+ * Class responsible for management of shader passes, associated with a device.
+ *
+ * @ignore
+ */
+class ShaderPass {
     /**
-     * Returns true if the shader pass is a forward shader pass.
+     * Allocated shader passes, map of a shader pass name to info.
      *
-     * @param {number} pass - The shader pass.
-     * @returns {boolean} - True if the pass is a forward shader pass.
+     * @type {Map<string, ShaderPassInfo>}
+     * @ignore
```

Comment:
AFAIK, if there is an `@ignore` on the class description, you don't need `@ignore` on anything in the class itself.

Reply:
yep fixed both, thanks